# Pandas

Pandas is an open source library and is built on top of NumPy. It is one of the most important libraries when it comes to data
analysis. It has an in-built visulaization features and allows to access data from various sources. 

We shall be covering following topics:
1) Series
2) DataFrames
3) Adding or dropping a column/ row
4) Row selection (loc functions)
5) Filtering data 
6) Missing Values
7) Groupby
8) Concatenation and Merging
9) Operations

### Series

Series are one dimensional objects like labelled array and can hold data of any data type (int, float, string, ...)

In [1]:
import pandas as pd
import numpy as np

#### Creating a series

Syntax - pandas.Series( data, index, dtype, copy)
except data all other arguments are optional

In [2]:
# Creating a simple series
pd.Series((1,2,3)) 

0    1
1    2
2    3
dtype: int64

In [3]:
# Creating series with lables
pd.Series([1,2,3], ['a','b','c']) 

a    1
b    2
c    3
dtype: int64

In [4]:
#Creating series with lables and data type
pd.Series([14,25,39], [0,1,2], float)

0    14.0
1    25.0
2    39.0
dtype: float64

In [5]:
# Creating a series where index is same as data
data = np.array([6,4,8,90])
pd.Series(data,data)

6      6
4      4
8      8
90    90
dtype: int32

In [6]:
# Scalar series object
pd.Series(23, [1,2,3,4])

1    23
2    23
3    23
4    23
dtype: int64

#### Lookups using Indexes

In [7]:
# Creating a sample series object
ser1 = pd.Series([34,45,789], ['james', 'tony', 'rex'])

In [8]:
ser1

james     34
tony      45
rex      789
dtype: int64

##### Lookup using index name/ label

In [9]:
# Fetching single elements by index
ser1['rex']

789

In [10]:
# Double square brackets to fetch more than one element
ser1[['rex', 'james']]

rex      789
james     34
dtype: int64

##### Lookup using index position

In [11]:
ser1[:2] # Includes 0th index and 1st index

james    34
tony     45
dtype: int64

In [12]:
ser1[:1] # 0th index only

james    34
dtype: int64

In [13]:
ser1[-1] # last element

789

##### Applying mathematical operations on Series

In [14]:
ser1 = pd.Series([32,3,2], ['k', 'p', 'r'])

In [15]:
ser2 = pd.Series([4,30,29, 12], ['k', 'p', 'r', 's'])

In [16]:
# Adding series objects. Indexes that are not present in one of the series will return NaN in the output
ser1+ser2

k    36.0
p    33.0
r    31.0
s     NaN
dtype: float64

### DataFrames

DataFrames store data in row x column format.
Syntax - pd.DataFrame( data, index, columns, dtype, copy)
where,
data - actual data
index - row labels
columns - headers
dtype - data type
copy - used for copying data, by default it is FALSE

Let us start by creating a simple DataFrame

In [507]:
# Here we are creating a DataFrame with five rows and four columns. Here index have been given custom names
df = pd.DataFrame(np.random.randn(5,4), index = [9,8,7,6,5], columns = ['a', 'b', 'c', 'd'])

In [508]:
df

a         b         c         d
9 -1.628955  0.544509 -0.113831  0.579082
8  0.036843 -0.993996  0.417418 -0.045647
7 -0.105751 -0.841282  1.263340  0.982564
6  0.613603 -0.307633 -0.067419  0.223460
5  2.057287  2.198455 -0.501855 -0.104383

If you dont specify index labels, it will by default start from 0. Let us try with an example

In [509]:
df = pd.DataFrame(np.random.randn(5,4), columns = ['a', 'b', 'c', 'd'])

In [510]:
df

a         b         c         d
0  0.897543 -1.044133 -1.887178 -1.375657
1  1.056311  0.873259  0.615070 -1.281747
2  0.663415 -0.772728  1.001137 -0.080061
3  0.371415 -0.638752  0.201842 -0.083424
4  0.520677  0.186007  0.265135  1.360632

Let us create a more meaningful DataFrame and do some operations on it

In [511]:
df = pd.DataFrame([[101, 'James', 50000, 'Chicago'], [201, 'Tom', 35000, 'New York'], [301, 'Starc', 20000, 'Dallas'], 
                  [401, 'Phil', 100000, 'Los Angeles'], [501, 'Rick', 42000, 'San Francisco']], 
                  columns = ['ID', 'Name', 'Income', 'City'])

In [512]:
df

ID   Name  Income           City
0  101  James   50000        Chicago
1  201    Tom   35000       New York
2  301  Starc   20000         Dallas
3  401   Phil  100000    Los Angeles
4  501   Rick   42000  San Francisco

The above dataframe can also be created in following manner

In [513]:
data = [[101, 'James', 50000, 'Chicago'], [201, 'Tom', 35000, 'New York'], [301, 'Starc', 20000, 'Dallas'], 
                  [401, 'Phil', 100000, 'Los Angeles'], [501, 'Rick', 42000, 'San Francisco']]

In [514]:
columns = ['ID', 'Name', 'Income', 'City']

In [515]:
df = pd.DataFrame(data, columns = columns)

In [516]:
df

ID   Name  Income           City
0  101  James   50000        Chicago
1  201    Tom   35000       New York
2  301  Starc   20000         Dallas
3  401   Phil  100000    Los Angeles
4  501   Rick   42000  San Francisco

Let us select a column and check the type

In [517]:
type(df['ID'])

pandas.core.series.Series

So we can see here that a single column is of series type, so DataFrames are nothing but series of columns joined together

#### Column Selection

There are two ways by which you can select a column in Pandas
Method 1: Similar to SQL df.column_name
          This method will enable you to pick only one column at a time. So refrain yourself from using this method
Method 2: Using square brckets df.['column_name']
          This is the most colmmonly method used for selecting columns. You can select single column or multiple columns using
          this method

Method 1: df.column_name

In [518]:
# Picking ID column
df.ID

0    101
1    201
2    301
3    401
4    501
Name: ID, dtype: int64

Method 2: df['column_name']

In [519]:
# Picking ID columns
df['ID']

0    101
1    201
2    301
3    401
4    501
Name: ID, dtype: int64

In [520]:
# Picking multiple columns
df[['ID', 'City']]

ID           City
0  101        Chicago
1  201       New York
2  301         Dallas
3  401    Los Angeles
4  501  San Francisco

Make sure that you use double square brackets will selecting more than one column, else it will throw an error

#### Adding a new column to your DataFrame

Normally column is added by applying conditions on existing columns. For example while preparing dataset for building a model you may want to create a new column by applying some formula on existing column. For example calculating annual income from monthly income. Let us try doing this here

In [521]:
# Getting annual income by multiplying monthly income by 12
df['Annual_Income'] = df['Income'] * 12

In [522]:
df

ID   Name  Income           City  Annual_Income
0  101  James   50000        Chicago         600000
1  201    Tom   35000       New York         420000
2  301  Starc   20000         Dallas         240000
3  401   Phil  100000    Los Angeles        1200000
4  501   Rick   42000  San Francisco         504000

Let us create one more column from Income

We will use here np.where function. This is exactly same as if() function in excel. Similar to if() function you can have nested
np.where() function in numpy

In [527]:
df['Income_Bin'] = np.where(df['Income'] <= 35000, 'a.<=35k',
                           (np.where(df['Income'] <= 50000, 'b.>35k-<=50k', 'c.>50k')))

In [528]:
df

ID   Name  Income           City  Annual_Income    Income_Bin
0  101  James   50000        Chicago         600000  b.>35k-<=50k
1  201    Tom   35000       New York         420000       a.<=35k
2  301  Starc   20000         Dallas         240000       a.<=35k
3  401   Phil  100000    Los Angeles        1200000        c.>50k
4  501   Rick   42000  San Francisco         504000  b.>35k-<=50k

There are many ways by which you can create such columns; one of the methods that is most commonly used are functions

#### Dropping a column from DataFrame

We will use df.drop() function. It can be used for dropping column as well as the row. Let us start with dropping a column

In [529]:
# Dropping column Income_Bin
df.drop(['Income_Bin'], axis = 1)

ID   Name  Income           City  Annual_Income
0  101  James   50000        Chicago         600000
1  201    Tom   35000       New York         420000
2  301  Starc   20000         Dallas         240000
3  401   Phil  100000    Los Angeles        1200000
4  501   Rick   42000  San Francisco         504000

Here axis = 1 means we are dropping a column. If you dont specify axis it will by default ir will drop row.
So if you are not specifying axis = 1 and trying to drop a column, it will try to find a row with that row-index label. If not
found it will throw an error

One More important point to note here is with above command it will not make changes in the original DataFrame; though the 
result displayed here does not contain Income_Bin. Let us try and see if Income_Bin has been dropped from original DataFrame or
not

In [530]:
df

ID   Name  Income           City  Annual_Income    Income_Bin
0  101  James   50000        Chicago         600000  b.>35k-<=50k
1  201    Tom   35000       New York         420000       a.<=35k
2  301  Starc   20000         Dallas         240000       a.<=35k
3  401   Phil  100000    Los Angeles        1200000        c.>50k
4  501   Rick   42000  San Francisco         504000  b.>35k-<=50k

So you can see here that original DataFrame has Income_Bin column. So in order to drop this column from original DataFrame, use
"inplace = True" as an argument in the drop function. By default it is False

In [531]:
df.drop(['Income_Bin'], axis = 1, inplace = True)

In [532]:
df

ID   Name  Income           City  Annual_Income
0  101  James   50000        Chicago         600000
1  201    Tom   35000       New York         420000
2  301  Starc   20000         Dallas         240000
3  401   Phil  100000    Los Angeles        1200000
4  501   Rick   42000  San Francisco         504000

Dropping mulitple columns from the dataframe

In [533]:
df.drop(['Income', 'City'], axis = 1, inplace = True)

In [534]:
df

ID   Name  Annual_Income
0  101  James         600000
1  201    Tom         420000
2  301  Starc         240000
3  401   Phil        1200000
4  501   Rick         504000

#### Row Selection

Rows can be selected based on following three methods:
    1) Based in index position or label
    2) By applying condition on single column or multiple columns
    3) Selecting randomly
Here we will discuss first two methods, third will be taken care in Machine Learning notes

Method 1 - Based on index position or label
Here we will see two functions - loc[] and iloc[]
Pandas also has ix[] function, but we wont be getting into its details as it has deprecated

##### First Function .loc[]
gets rows (or columns) with particular labels from the index.

We already have one DataFrame df having default index 0,1,2,3... Let us create one more DataFrame with custom row index labels

In [66]:
df

ID   Name  Income           City
0  101  James   50000        Chicago
1  201    Tom   35000       New York
2  301  Starc   20000         Dallas
3  401   Phil  100000    Los Angeles
4  501   Rick   42000  San Francisco

In [538]:
data = [[101, 'James', 50000, 'Chicago'], [201, 'Tom', 35000, 'New York'], [301, 'Starc', 20000, 'Dallas'], 
                  [401, 'Phil', 100000, 'Los Angeles'], [501, 'Rick', 42000, 'San Francisco']]

In [539]:
columns = ['ID', 'Name', 'Income', 'City']

In [540]:
df1 = pd.DataFrame(data, columns = columns, index = ('X', 'T', 'R', 'F', 'K'))

In [541]:
df1

ID   Name  Income           City
X  101  James   50000        Chicago
T  201    Tom   35000       New York
R  301  Starc   20000         Dallas
F  401   Phil  100000    Los Angeles
K  501   Rick   42000  San Francisco

So now we have two DataFrames, one with default row lables and other with user defined row labels. Let us use loc function and 
understand how it works in row selection

##### Part a - Selecting rows based on rows label

In [542]:
# Getting a row based on the row label
df.loc[1]

ID                  201
Name                Tom
Annual_Income    420000
Name: 1, dtype: object

In [543]:
# Getting a row based on the row label
df1.loc['R']

ID           301
Name       Starc
Income     20000
City      Dallas
Name: R, dtype: object

In [544]:
# Getting rows upto a lable
df.loc[1:3] # Here it will select rows with label 1,2 and 3

ID   Name  Annual_Income
1  201    Tom         420000
2  301  Starc         240000
3  401   Phil        1200000

In [545]:
# Getting rows upto a lable
df1.loc['T': 'F'] # Here it will select rows with label T,R and F

ID   Name  Income         City
T  201    Tom   35000     New York
R  301  Starc   20000       Dallas
F  401   Phil  100000  Los Angeles

###### Part b - Selecting rows and columns together by rows and columns label

.loc[] function can also be used to select rows and columns together, it is a sort of subset of data. Consider the below 
example

In [546]:
# Selection 3 rows and 3 columns
df1.loc[['T', 'R', 'F'], ['ID', 'Name', 'Income']]

ID   Name  Income
T  201    Tom   35000
R  301  Starc   20000
F  401   Phil  100000

In [547]:
# One more way of selecting rows and columns
df1.loc['X':'F', 'ID':'Income']

ID   Name  Income
X  101  James   50000
T  201    Tom   35000
R  301  Starc   20000
F  401   Phil  100000

###### Part c - Selection rows and columns based on condition

This is similar to "where" function of SQL where you filter the data based on a condition applied on a single column or multiple columns. This fuunctionaility of .loc[] function is very useful for data preperation of predictive model development

In [548]:
# Create a subset with Name as James
df1.loc[df1['Name'] == 'James']

ID   Name  Income     City
X  101  James   50000  Chicago

In [549]:
# Select records with Income >= 50000
df1.loc[df1['Income'] >= 50000]

ID   Name  Income         City
X  101  James   50000      Chicago
F  401   Phil  100000  Los Angeles

In [550]:
# Select records with Income >= 20000 and Name as Phil
df1.loc[(df1['Income'] >= 20000) & (df1['Name'] == 'Phil')]

ID  Name  Income         City
F  401  Phil  100000  Los Angeles

##### Second function .iloc[]
gets rows (or columns) at particular positions in the index (so it only takes integers).

.iloc[] works based on the position of the rows and columns. By default when you create a DataFrame the rows and columns index
starts from 0. Do notice here that it takes only integer as an argument, unlike .loc[] which takes labels as an argument.
Below format will be used while applying .iloc[] function

               Column
           
Row   0     1     2     3     4 
        
 0   22    34    45    34     1
        
 1   35    56    13    98    45
        
 2   90    76    21    67    31

Let us see by trying few examples

In [551]:
# Applying .iloc[] on df dataFrame
df.iloc[0]

ID                  101
Name              James
Annual_Income    600000
Name: 0, dtype: object

The above query has returned 0th row position of data in a series type. Remember that if single record is fetched then it
will be returned as a series type

In [552]:
# Fetching multiple records from df by applying .iloc[] on df dataFrame
df.iloc[:3]

ID   Name  Annual_Income
0  101  James         600000
1  201    Tom         420000
2  301  Starc         240000

So if you notice here :x will alwasy return records till x-1 row indexes

In [553]:
# Return records with 1st index to 3rd index (row)
df.iloc[1:4]

ID   Name  Annual_Income
1  201    Tom         420000
2  301  Starc         240000
3  401   Phil        1200000

In [554]:
# Fetching selected rows and columns
df.iloc[1:4, 1:3]
# So here 1st, 2nd and 3rd index rows are picked and 1st and 2nd index columns are picked

Name  Annual_Income
1    Tom         420000
2  Starc         240000
3   Phil        1200000

Let us apply this to df1 DataFrame which has user labeled row indexes

In [556]:
df1

ID   Name  Income           City
X  101  James   50000        Chicago
T  201    Tom   35000       New York
R  301  Starc   20000         Dallas
F  401   Phil  100000    Los Angeles
K  501   Rick   42000  San Francisco

In [557]:
# Fetching 0th, 1st and 2nd index row
df1.iloc[:3]

ID   Name  Income      City
X  101  James   50000   Chicago
T  201    Tom   35000  New York
R  301  Starc   20000    Dallas

In [558]:
# Fetching 1st and 2nd row index and 0th, 1st and 2nd column index
df1.iloc[1:3, :3]

ID   Name  Income
T  201    Tom   35000
R  301  Starc   20000

One important difference between .loc[] and .iloc[] is conditional filters. By using .loc[] you filter the data based on a 
specific column condition, where as in .iloc[] it is not possible. Generally for data preperation .loc[] is used

#### Filtering data using conditions

You can use .loc[] function for filtering the DataFrame and create a subset based on a condition. But there are many ways by 
which you can do this. let us learn one more technique on how you can achieve this

In [559]:
# Consider following data Frame:
data = [[101, 'James', 50000, 'Chicago'], [201, 'Tom', 35000, 'New York'], [301, 'Starc', 20000, 'Dallas'], 
        [401, 'Phil', 100000, 'Los Angeles'], [501, 'Rick', 42000, 'San Francisco'], [601, 'Harry', 135000, 'Vegas'],
        [701, 'Jim', 25000, 'Los Angeles'], [801, 'Ricky', 75000, 'Chicago']]

In [560]:
columns = ['ID', 'Name', 'Income', 'City']

In [561]:
df = pd.DataFrame(data, columns = columns)

In [562]:
df

ID   Name  Income           City
0  101  James   50000        Chicago
1  201    Tom   35000       New York
2  301  Starc   20000         Dallas
3  401   Phil  100000    Los Angeles
4  501   Rick   42000  San Francisco
5  601  Harry  135000          Vegas
6  701    Jim   25000    Los Angeles
7  801  Ricky   75000        Chicago

In [563]:
# Let us create a DataFrame from existing one where Income > 25000
df[(df['Income'] > 25000)]

ID   Name  Income           City
0  101  James   50000        Chicago
1  201    Tom   35000       New York
3  401   Phil  100000    Los Angeles
4  501   Rick   42000  San Francisco
5  601  Harry  135000          Vegas
7  801  Ricky   75000        Chicago

So here we have all the records where Income >= 25000

In [564]:
# Let us run the same query above, but selecting only few columns
df[(df['Income'] > 25000)][['City']]

City
0        Chicago
1       New York
3    Los Angeles
4  San Francisco
5          Vegas
7        Chicago

So we just have to specify names of the columns after specifying the conditions. Just remember that while pulling out a single 
column put column name in double square brackets to have output as a DataFrame. Else it will be a series. Let us try this out

In [565]:
df[(df['Income'] > 25000)]['City']

0          Chicago
1         New York
3      Los Angeles
4    San Francisco
5            Vegas
7          Chicago
Name: City, dtype: object

In [566]:
# Slecting multiple columns
df[(df['Income'] > 25000)][['City', 'Name']]

City   Name
0        Chicago  James
1       New York    Tom
3    Los Angeles   Phil
4  San Francisco   Rick
5          Vegas  Harry
7        Chicago  Ricky

In [567]:
# Applying Multiple conditions (using and function)
df[(df['Income'] > 25000) & (df['City'] == 'Chicago')][['City', 'ID', 'Income']]

City   ID  Income
0  Chicago  101   50000
7  Chicago  801   75000

In [568]:
# Applying Multiple conditions (using or function)
df[(df['Income'] > 25000) | (df['City'] == 'Chicago')][['City', 'ID', 'Income']]

City   ID  Income
0        Chicago  101   50000
1       New York  201   35000
3    Los Angeles  401  100000
4  San Francisco  501   42000
5          Vegas  601  135000
7        Chicago  801   75000

#### Reindexing
Multiple tasks can be acheived using reindexing. For example we can reset the existing indexes to default one or we can make an existing column as an index of the DataFrame. Let us have a look at these one by one

In [569]:
# Considering df1 DataFrame
df1

ID   Name  Income           City
X  101  James   50000        Chicago
T  201    Tom   35000       New York
R  301  Starc   20000         Dallas
F  401   Phil  100000    Los Angeles
K  501   Rick   42000  San Francisco

In [570]:
# Reset index to default 
df1.reset_index()

index   ID   Name  Income           City
0     X  101  James   50000        Chicago
1     T  201    Tom   35000       New York
2     R  301  Starc   20000         Dallas
3     F  401   Phil  100000    Los Angeles
4     K  501   Rick   42000  San Francisco

So we see here that original index (X, T, R ...) have been converted to a column and a default index has been set. One more important point to note here is that the original DataFrame has not changed at all. We would have to use 'inplace = True' to make that change happen in original DataFrame. Let us quickly try this out`

In [571]:
df2 = df1 # I dont want to make changes in the original DataFrame

In [572]:
df2

ID   Name  Income           City
X  101  James   50000        Chicago
T  201    Tom   35000       New York
R  301  Starc   20000         Dallas
F  401   Phil  100000    Los Angeles
K  501   Rick   42000  San Francisco

In [573]:
# Resetting the indexing with inplace = True argument
df2.reset_index(inplace = True)

In [574]:
df2

index   ID   Name  Income           City
0     X  101  James   50000        Chicago
1     T  201    Tom   35000       New York
2     R  301  Starc   20000         Dallas
3     F  401   Phil  100000    Los Angeles
4     K  501   Rick   42000  San Francisco

Let us use one more argument of reset_index function which will drop the index column

In [576]:
# Let us use below DataFrame to check this functionality
df

ID   Name  Income           City
0  101  James   50000        Chicago
1  201    Tom   35000       New York
2  301  Starc   20000         Dallas
3  401   Phil  100000    Los Angeles
4  501   Rick   42000  San Francisco
5  601  Harry  135000          Vegas
6  701    Jim   25000    Los Angeles
7  801  Ricky   75000        Chicago

In [577]:
df.reset_index(inplace = True, drop = True)

In [578]:
df

ID   Name  Income           City
0  101  James   50000        Chicago
1  201    Tom   35000       New York
2  301  Starc   20000         Dallas
3  401   Phil  100000    Los Angeles
4  501   Rick   42000  San Francisco
5  601  Harry  135000          Vegas
6  701    Jim   25000    Los Angeles
7  801  Ricky   75000        Chicago

So we can see here that index has been reset and old index column has also been dropped

Let us try to assign an existing column as an index. We will use the above DataFrame.

In [579]:
# Setting City as an index
df.set_index('City')

ID   Name  Income
City                             
Chicago        101  James   50000
New York       201    Tom   35000
Dallas         301  Starc   20000
Los Angeles    401   Phil  100000
San Francisco  501   Rick   42000
Vegas          601  Harry  135000
Los Angeles    701    Jim   25000
Chicago        801  Ricky   75000

To make this change happen in original DataFrame, we will have to use inplace = True 

#### Missing Values

Missing values are very important in any Data Preperation process. Treating missing values invloves a lot of steps - Identifying
missing values and imputing values. Let us create a DataFrame where we have Missing Values

In [580]:
# Consider following data Frame:
data = [[101, 'James', 50000, 'Chicago', 27], [201, 'Tom', 35000, 'New York', 33], [301, 'Starc', 20000, 'Dallas', 21], 
        [401, 'Phil', 100000, np.nan, 40], [501, 'Rick', np.nan, 'San Francisco', np.nan], 
        [601, 'Harry', 135000, 'Vegas', 44], [701, 'Jim', 25000, 'Los Angeles', 24], [801, 'Ricky', 75000, 'Chicago', np.nan]]

In [581]:
columns = ['Id', 'Names', 'Income', 'City', 'Age']

In [582]:
df = pd.DataFrame(data, columns = columns)

In [583]:
df

Id  Names    Income           City   Age
0  101  James   50000.0        Chicago  27.0
1  201    Tom   35000.0       New York  33.0
2  301  Starc   20000.0         Dallas  21.0
3  401   Phil  100000.0            NaN  40.0
4  501   Rick       NaN  San Francisco   NaN
5  601  Harry  135000.0          Vegas  44.0
6  701    Jim   25000.0    Los Angeles  24.0
7  801  Ricky   75000.0        Chicago   NaN

##### Identifying Null/ Missing Values

In [584]:
df.isnull()

Id  Names  Income   City    Age
0  False  False   False  False  False
1  False  False   False  False  False
2  False  False   False  False  False
3  False  False   False   True  False
4  False  False    True  False   True
5  False  False   False  False  False
6  False  False   False  False  False
7  False  False   False  False   True

Code df.isnull() will check each and every cell and return TRUE or FALSE depending on the condition. It will not be a good idea 
to run this code over a large DataFrame as it will be difficult to analyze which cell has Missing Value. So let us try to get
overall count of Missing Values, count of Missing Values in each column ...

In [585]:
# Counting Missing values in each column
df.isnull().sum()

Id        0
Names     0
Income    1
City      1
Age       2
dtype: int64

In [586]:
# Counting total number of Missing values in the entire DataFrame
df.isnull().sum().sum()

4

So with above two codes we could easily understand number of missing values present in the entire DataFrame and each column.
This gives us more understanding on the data availability for each column also a starting point

Let us now go one step further by calculating % of Missing values in each column. This is very important as count does not give
a clear picture on data availability, especially when you are comparing missing value distribution of each column. You might
be interested in rejecting in a column from your analysis depending on the % of data availability

In [587]:
# Calculating % of Missing values in each column
df.isnull().sum() * 100/len(df)

Id         0.0
Names      0.0
Income    12.5
City      12.5
Age       25.0
dtype: float64

If you notice, above output is a series and can be easily converted to a DataFrame. Let us do it

In [588]:
x = pd.DataFrame(df.isnull().sum() * 100/len(df))

In [589]:
x

0
Id       0.0
Names    0.0
Income  12.5
City    12.5
Age     25.0

In [590]:
x.reset_index()

index     0
0      Id   0.0
1   Names   0.0
2  Income  12.5
3    City  12.5
4     Age  25.0

##### Dropping Null Values

It is important to understand on what basis do we drop a column or a row. For example it will be meaningful to drop a row
which has most of its data missing or drop a column which has maximum number of data points missing. Let us understand on how to
drop Missing values

In [591]:
# Drop records/ rows having at least one missing value
df.dropna()

Id  Names    Income         City   Age
0  101  James   50000.0      Chicago  27.0
1  201    Tom   35000.0     New York  33.0
2  301  Starc   20000.0       Dallas  21.0
5  601  Harry  135000.0        Vegas  44.0
6  701    Jim   25000.0  Los Angeles  24.0

In [592]:
# Drops columns having at least one missing value
df.dropna(axis = 1)

Id  Names
0  101  James
1  201    Tom
2  301  Starc
3  401   Phil
4  501   Rick
5  601  Harry
6  701    Jim
7  801  Ricky

So argument axis by default is set to 0 which will delete rows. Agument axis = 1 will drop columns with at least one missing
value. Now it will not be fair to drop a row or column with even one missing value. It will be good to analyse the missing 
value distribution across dataset and then decide on deletion of rows/ columns.

In [593]:
# Refering to the 'x' data that we created
x

0
Id       0.0
Names    0.0
Income  12.5
City    12.5
Age     25.0

Here we see that Age is not present for 25% of the records, so let us try dropping Age from the DataFrame

In [594]:
# Dropping column Age
df.drop(['Age'], axis = 1)

Id  Names    Income           City
0  101  James   50000.0        Chicago
1  201    Tom   35000.0       New York
2  301  Starc   20000.0         Dallas
3  401   Phil  100000.0            NaN
4  501   Rick       NaN  San Francisco
5  601  Harry  135000.0          Vegas
6  701    Jim   25000.0    Los Angeles
7  801  Ricky   75000.0        Chicago

Now we have seen how we can drop columns. For deciding on the logic for deleting rows/ records we need to check for how many 
variables the data is missing. For example for a particular record data might be missing for all the variables or for most of
the variables. In the above example we have five columns, so let us delete records where data is missing for 2 or more variables.
Now let us check on how we can delete rows/ records

In [596]:
# Dropping records where data is available for at least 4 of the variables
df.dropna(thresh = 4)

Id  Names    Income         City   Age
0  101  James   50000.0      Chicago  27.0
1  201    Tom   35000.0     New York  33.0
2  301  Starc   20000.0       Dallas  21.0
3  401   Phil  100000.0          NaN  40.0
5  601  Harry  135000.0        Vegas  44.0
6  701    Jim   25000.0  Los Angeles  24.0
7  801  Ricky   75000.0      Chicago   NaN

So row with index 4 has been deleted from the DataFrame as it has data vailable only for 3 variables.

##### Imputing Missing Values

Imputing Null values is nothing but Missing Value Treatment. There are lot of techniques by which you can do this. You can
impute the missing value by mean, median, mode or use regression method. Let us discuss impute missing values by mean, median or
mode

In [597]:
# Imputing missing values of age by mean of age (mean of age will be taken wherever age is present)
df['Age'].fillna(df['Age'].mean())

0    27.0
1    33.0
2    21.0
3    40.0
4    31.5
5    44.0
6    24.0
7    31.5
Name: Age, dtype: float64

So mean age is 31.5 which has been put in place wherever age was missing

In [598]:
# Imputing missing values of age by median of age (median of age will be taken wherever age is present)
df['Age'].fillna(df['Age'].median())

0    27.0
1    33.0
2    21.0
3    40.0
4    30.0
5    44.0
6    24.0
7    30.0
Name: Age, dtype: float64

#### Groupby

This is very important for summarizing any data. It is used to group data with respect to column using aggregate function like
sum or count. People familiar to SQL will have a very good understanding of this. Let us create a dataset on which we can apply
groupby function

In [599]:
# Consider following data Frame:
data = [[101, 'James', 50000, 'Chicago', 27, 'HR', 700000, 'Y'], [201, 'Tom', 35000, 'New York', 33, 'Analytics', 550000, 'Y'], 
        [301, 'Starc', 20000, 'Dallas', 21, 'IT', 125000, 'N'], [401, 'Phil', 100000, 'Dallas', 40, 'HR', 800000, 'Y'], 
        [501, 'Rick', np.nan, 'San Francisco', np.nan, 'IT', 75000, 'Y'], 
        [601, 'Harry', 135000, 'Vegas', 44, 'Marketing', 900000, 'N'], 
        [701, 'Jim', 25000, 'Los Angeles', 24, 'HR', 350000, 'Y'], 
        [801, 'Ricky', 75000, 'Chicago', np.nan, 'Analytics', 450000, 'N'],
        [901, 'Fin', 85000, 'Vegas', 51, 'Marketing', 600000, 'N'], [1001, 'Grant', 61500, 'New York', 39, 'IT', 75000, 'N'], 
        [1002, 'Cooper', 31500, 'Dallas', 27, 'IT', 250000, 'N'], 
        [1003, 'Brand', 40500, 'San Francisco', 49, 'Marketing', 50000, 'N']]

In [600]:
columns = ['ID', 'Name', 'Income', 'City', 'Age', 'Department', 'Sales', 'Response']

In [601]:
df = pd.DataFrame(data, columns = columns)

In [602]:
df

ID    Name    Income           City   Age Department   Sales Response
0    101   James   50000.0        Chicago  27.0         HR  700000        Y
1    201     Tom   35000.0       New York  33.0  Analytics  550000        Y
2    301   Starc   20000.0         Dallas  21.0         IT  125000        N
3    401    Phil  100000.0         Dallas  40.0         HR  800000        Y
4    501    Rick       NaN  San Francisco   NaN         IT   75000        Y
5    601   Harry  135000.0          Vegas  44.0  Marketing  900000        N
6    701     Jim   25000.0    Los Angeles  24.0         HR  350000        Y
7    801   Ricky   75000.0        Chicago   NaN  Analytics  450000        N
8    901     Fin   85000.0          Vegas  51.0  Marketing  600000        N
9   1001   Grant   61500.0       New York  39.0         IT   75000        N
10  1002  Cooper   31500.0         Dallas  27.0         IT  250000        N
11  1003   Brand   40500.0  San Francisco  49.0  Marketing   50000        N

Group by process is divided in two 3 steps:
1) Identify variables by which we would like to summarize the data
2) Create a groupby object and store it in a variable
3) Now summarize the data using aggregate functions
Let us get started now

##### Groupby Single column/ variable

Step 1 - Identify variable - City
Step 2 - Create a groupby object and store it in a variable

In [603]:
by_city = df.groupby('City')

Step 3 - Summarize the data using aggregate functions

In [604]:
# Let us check total sales across cities
by_city['Sales'].sum()

City
Chicago          1150000
Dallas           1175000
Los Angeles       350000
New York          625000
San Francisco     125000
Vegas            1500000
Name: Sales, dtype: int64

In [605]:
# Counting employees across cities
by_city['ID'].count()

City
Chicago          2
Dallas           3
Los Angeles      1
New York         2
San Francisco    2
Vegas            2
Name: ID, dtype: int64

In [606]:
# Getting max age across cities
by_city['Age'].max()

City
Chicago          27.0
Dallas           40.0
Los Angeles      24.0
New York         39.0
San Francisco    49.0
Vegas            51.0
Name: Age, dtype: float64

In [607]:
by_city['Income', 'Sales'].sum()

Income    Sales
City                            
Chicago        125000.0  1150000
Dallas         151500.0  1175000
Los Angeles     25000.0   350000
New York        96500.0   625000
San Francisco   40500.0   125000
Vegas          220000.0  1500000

In [608]:
# Multiple aggregate function
by_city['Sales'].agg([np.mean, np.sum, np.min, np.max, np.std])

mean      sum    amin    amax            std
City                                                                
Chicago        575000.000000  1150000  450000  700000  176776.695297
Dallas         391666.666667  1175000  125000  800000  359107.690440
Los Angeles    350000.000000   350000  350000  350000            NaN
New York       312500.000000   625000   75000  550000  335875.721064
San Francisco   62500.000000   125000   50000   75000   17677.669530
Vegas          750000.000000  1500000  600000  900000  212132.034356

##### Groupby Multiple columns/ variable

In [609]:
# Groupby City and Department
by_city_dept = df.groupby(['City', 'Department', 'Response'])

In [610]:
# Using sum() as an aggregate function
pd.DataFrame(by_city_dept['ID'].count())

ID
City          Department Response    
Chicago       Analytics  N          1
              HR         Y          1
Dallas        HR         Y          1
              IT         N          2
Los Angeles   HR         Y          1
New York      Analytics  Y          1
              IT         N          1
San Francisco IT         Y          1
              Marketing  N          1
Vegas         Marketing  N          2

An important observation here, while executing aggregate functions it does not consider Missing values. They are skipped for all
the computations

#### Concatenation and Merging

##### Concatenation

Concatenation combines two DataFrames. Function used:
pd.concat(objs, axis=0, join='outer', join_axes=None, ignore_index=False)
where,
objs - DataFrames
axis - Along row or column
join - Outer or Inner, Outer for union and inner for intersection.

Let us get started by creating two DataFrames which we can use for understanding concat function

In [611]:
# Consider following data Frame:
data1 = [[101, 'James', 50000, 'Chicago', 27, 'HR', 700000, 'Y'], 
         [201, 'Tom', 35000, 'New York', 33, 'Analytics', 550000, 'Y'], 
         [901, 'Fin', 85000, 'Vegas', 51, 'Marketing', 600000, 'N'], [1001, 'Grant', 61500, 'New York', 39, 'IT', 75000, 'N'], 
         [1002, 'Cooper', 31500, 'Dallas', 27, 'IT', 250000, 'N'], 
         [1003, 'Brand', 40500, 'San Francisco', 49, 'Marketing', 50000, 'N']]

In [612]:
columns1 = ['ID', 'Name', 'Income', 'City', 'Age', 'Department', 'Sales', 'Response']

In [613]:
df1 = pd.DataFrame(data1, columns = columns1)

In [614]:
df1

ID    Name  Income           City  Age Department   Sales Response
0   101   James   50000        Chicago   27         HR  700000        Y
1   201     Tom   35000       New York   33  Analytics  550000        Y
2   901     Fin   85000          Vegas   51  Marketing  600000        N
3  1001   Grant   61500       New York   39         IT   75000        N
4  1002  Cooper   31500         Dallas   27         IT  250000        N
5  1003   Brand   40500  San Francisco   49  Marketing   50000        N

In [615]:
data2 = [[1101, 'James', 50000, 'Chicago', 27, 'HR', 700000, 'Y'], 
         [2101, 'Tom', 35000, 'New York', 33, 'Analytics', 550000, 'Y'], 
         [9101, 'Fin', 85000, 'Vegas', 51, 'Marketing', 600000, 'N'], 
         [11001, 'Grant', 61500, 'New York', 39, 'IT', 75000, 'N'], 
         [1002, 'Cooper', 31500, 'Dallas', 27, 'IT', 250000, 'N'], 
         [11003, 'Brand', 40500, 'San Francisco', 49, 'Marketing', 50000, 'N']]

In [616]:
columns2 = ['ID', 'Name', 'Income', 'City', 'Age', 'Department', 'Sales', 'Response']

In [617]:
df2 = pd.DataFrame(data2, columns = columns2)

In [618]:
df2

ID    Name  Income           City  Age Department   Sales Response
0   1101   James   50000        Chicago   27         HR  700000        Y
1   2101     Tom   35000       New York   33  Analytics  550000        Y
2   9101     Fin   85000          Vegas   51  Marketing  600000        N
3  11001   Grant   61500       New York   39         IT   75000        N
4   1002  Cooper   31500         Dallas   27         IT  250000        N
5  11003   Brand   40500  San Francisco   49  Marketing   50000        N

In [619]:
# Concatenate df1 and df2 and let us observe the result (by default it will take axis = 0)
pd.concat([df1,df2])

ID    Name  Income           City  Age Department   Sales Response
0    101   James   50000        Chicago   27         HR  700000        Y
1    201     Tom   35000       New York   33  Analytics  550000        Y
2    901     Fin   85000          Vegas   51  Marketing  600000        N
3   1001   Grant   61500       New York   39         IT   75000        N
4   1002  Cooper   31500         Dallas   27         IT  250000        N
5   1003   Brand   40500  San Francisco   49  Marketing   50000        N
0   1101   James   50000        Chicago   27         HR  700000        Y
1   2101     Tom   35000       New York   33  Analytics  550000        Y
2   9101     Fin   85000          Vegas   51  Marketing  600000        N
3  11001   Grant   61500       New York   39         IT   75000        N
4   1002  Cooper   31500         Dallas   27         IT  250000        N
5  11003   Brand   40500  San Francisco   49  Marketing   50000        N

So the DataFrames along with indexes have got appended. We can reset the index of the resultant DataFrame by using "ignore_index" argument

In [620]:
pd.concat([df1, df2], ignore_index = True)

ID    Name  Income           City  Age Department   Sales Response
0     101   James   50000        Chicago   27         HR  700000        Y
1     201     Tom   35000       New York   33  Analytics  550000        Y
2     901     Fin   85000          Vegas   51  Marketing  600000        N
3    1001   Grant   61500       New York   39         IT   75000        N
4    1002  Cooper   31500         Dallas   27         IT  250000        N
5    1003   Brand   40500  San Francisco   49  Marketing   50000        N
6    1101   James   50000        Chicago   27         HR  700000        Y
7    2101     Tom   35000       New York   33  Analytics  550000        Y
8    9101     Fin   85000          Vegas   51  Marketing  600000        N
9   11001   Grant   61500       New York   39         IT   75000        N
10   1002  Cooper   31500         Dallas   27         IT  250000        N
11  11003   Brand   40500  San Francisco   49  Marketing   50000        N

Let us focus on one very important aspect of concat. In SQL we have heard of function called as UNION. It appends two 
datasets one below the other by making sure that records are unique. On the other hand UNION ALL takes duplicate records into the account.So The above resultant DataFrame is nothing but UNION ALL outcome. Let us see how we can get UNION output of SQL in 
Python

In [621]:
pd.concat([df1, df2], ignore_index=True).drop_duplicates().reset_index()

index     ID    Name  Income           City  Age Department   Sales  \
0       0    101   James   50000        Chicago   27         HR  700000   
1       1    201     Tom   35000       New York   33  Analytics  550000   
2       2    901     Fin   85000          Vegas   51  Marketing  600000   
3       3   1001   Grant   61500       New York   39         IT   75000   
4       4   1002  Cooper   31500         Dallas   27         IT  250000   
5       5   1003   Brand   40500  San Francisco   49  Marketing   50000   
6       6   1101   James   50000        Chicago   27         HR  700000   
7       7   2101     Tom   35000       New York   33  Analytics  550000   
8       8   9101     Fin   85000          Vegas   51  Marketing  600000   
9       9  11001   Grant   61500       New York   39         IT   75000   
10     11  11003   Brand   40500  San Francisco   49  Marketing   50000   

   Response  
0         Y  
1         Y  
2         N  
3         N  
4         N  
5         N  
6         Y  
7         Y  
8         N  
9         N  
10        N

So in the above output the duplicate record issue has been taken care of and also we have reset the index. This is very useful
for the data preperation process.

You can also concatenate DataFrames along columns by passing argument axis = 1. Let us see this with an example

In [622]:
pd.concat([df1, df2], axis = 1)

ID    Name  Income           City  Age Department   Sales Response  \
0   101   James   50000        Chicago   27         HR  700000        Y   
1   201     Tom   35000       New York   33  Analytics  550000        Y   
2   901     Fin   85000          Vegas   51  Marketing  600000        N   
3  1001   Grant   61500       New York   39         IT   75000        N   
4  1002  Cooper   31500         Dallas   27         IT  250000        N   
5  1003   Brand   40500  San Francisco   49  Marketing   50000        N   

      ID    Name  Income           City  Age Department   Sales Response  
0   1101   James   50000        Chicago   27         HR  700000        Y  
1   2101     Tom   35000       New York   33  Analytics  550000        Y  
2   9101     Fin   85000          Vegas   51  Marketing  600000        N  
3  11001   Grant   61500       New York   39         IT   75000        N  
4   1002  Cooper   31500         Dallas   27         IT  250000        N  
5  11003   Brand   40500  San Francisco   49  Marketing   50000        N

It matches the row indexes and then concatenates the DataFrames. In case indexes are not matching then it will fill values by 
NaN

##### Merging/ Joining

Very important concept and same as joing in SQL. It will be good to understand this in detail as Data Preperation might require
you to work on multiple Datasets having common primary key. Let us get started:

Syntax - 
pd.merge(left, right, how='inner', on=None, left_on=None, right_on=None, left_index=False, right_index=False, sort=True)
where,
left/ right - DataFrames
on - Column on which DataFrames are to be joined. Must be present in both the DataFrames
left_on - Column from left DataFrame used as key for joining
right_on - Column from right DataFrame used as key for joining
left_index - Use left DataFrame index for joining
right_index - Use right DataFrame index for joining
sort - Sort the result in lexicographical order

In [ ]:
Let us get started by creating two DataFrames which we can use for understanding merge function

In [463]:
# Consider following data Frame:
data1 =  [[101, 'James', 50000, 'Chicago', 27, 'HR', 700000, 'Y'], 
         [201, 'Tom', 35000, 'New York', 33, 'Analytics', 550000, 'Y'], 
         [901, 'Fin', 85000, 'Vegas', 51, 'Marketing', 600000, 'N'], 
         [1001, 'Grant', 61500, 'New York', 39, 'IT', 75000, 'N'], 
         [1002, 'Cooper', 31500, 'Dallas', 27, 'IT', 250000, 'N'], 
         [1003, 'Brand', 40500, 'San Francisco', 49, 'Marketing', 50000, 'N']]

In [464]:
columns1 = ['ID', 'Name', 'Income', 'City', 'Age', 'Department', 'Sales', 'Response']

In [465]:
df1 = pd.DataFrame(data1, columns = columns1)

In [466]:
df1

ID    Name  Income           City  Age Department   Sales Response
0   101   James   50000        Chicago   27         HR  700000        Y
1   201     Tom   35000       New York   33  Analytics  550000        Y
2   901     Fin   85000          Vegas   51  Marketing  600000        N
3  1001   Grant   61500       New York   39         IT   75000        N
4  1002  Cooper   31500         Dallas   27         IT  250000        N
5  1003   Brand   40500  San Francisco   49  Marketing   50000        N

In [467]:
# Consider following data Frame:
data2 = [[101, 'Married', 'PG'], [201, 'Single', 'Graduate'], [901, 'Married', 'HSC'], [1001, 'Single', 'Graduate'], 
        [1002, 'Married', 'Graduate'], [1003, 'Married', 'PG']]

In [468]:
columns2 = ['ID', 'Marital_Status', 'Education']

In [469]:
df2 = pd.DataFrame(data = data2, columns = columns2)

In [470]:
df2

ID Marital_Status Education
0   101        Married        PG
1   201         Single  Graduate
2   901        Married       HSC
3  1001         Single  Graduate
4  1002        Married  Graduate
5  1003        Married        PG

In [472]:
# Inner join to get all the columns of df1 and df2 in the same table
pd.merge(df1,df2, how = 'inner', on = 'ID')

ID    Name  Income           City  Age Department   Sales Response  \
0   101   James   50000        Chicago   27         HR  700000        Y   
1   201     Tom   35000       New York   33  Analytics  550000        Y   
2   901     Fin   85000          Vegas   51  Marketing  600000        N   
3  1001   Grant   61500       New York   39         IT   75000        N   
4  1002  Cooper   31500         Dallas   27         IT  250000        N   
5  1003   Brand   40500  San Francisco   49  Marketing   50000        N   

  Marital_Status Education  
0        Married        PG  
1         Single  Graduate  
2        Married       HSC  
3         Single  Graduate  
4        Married  Graduate  
5        Married        PG

In [473]:
# Left join to get all the columns of df1 and df2 in the same table
pd.merge(df1,df2, how = 'left', on = 'ID')

ID    Name  Income           City  Age Department   Sales Response  \
0   101   James   50000        Chicago   27         HR  700000        Y   
1   201     Tom   35000       New York   33  Analytics  550000        Y   
2   901     Fin   85000          Vegas   51  Marketing  600000        N   
3  1001   Grant   61500       New York   39         IT   75000        N   
4  1002  Cooper   31500         Dallas   27         IT  250000        N   
5  1003   Brand   40500  San Francisco   49  Marketing   50000        N   

  Marital_Status Education  
0        Married        PG  
1         Single  Graduate  
2        Married       HSC  
3         Single  Graduate  
4        Married  Graduate  
5        Married        PG

In [474]:
# Right join to get all the columns of df1 and df2 in the same table
pd.merge(df1,df2, how = 'right', on = 'ID')

ID    Name  Income           City  Age Department   Sales Response  \
0   101   James   50000        Chicago   27         HR  700000        Y   
1   201     Tom   35000       New York   33  Analytics  550000        Y   
2   901     Fin   85000          Vegas   51  Marketing  600000        N   
3  1001   Grant   61500       New York   39         IT   75000        N   
4  1002  Cooper   31500         Dallas   27         IT  250000        N   
5  1003   Brand   40500  San Francisco   49  Marketing   50000        N   

  Marital_Status Education  
0        Married        PG  
1         Single  Graduate  
2        Married       HSC  
3         Single  Graduate  
4        Married  Graduate  
5        Married        PG

In [476]:
# Selecting few columns from df1 and all from df2
pd.merge(df1[['ID', 'Name', 'Age']], df2, how = 'left', on = 'ID')

ID    Name  Age Marital_Status Education
0   101   James   27        Married        PG
1   201     Tom   33         Single  Graduate
2   901     Fin   51        Married       HSC
3  1001   Grant   39         Single  Graduate
4  1002  Cooper   27        Married  Graduate
5  1003   Brand   49        Married        PG

Like in SQL in Pandas also tables can be joined based on two columns. Just add the other column in "on" argument in double square brackets

#### Operations

In [623]:
# We would be using following DataFrame for understanding few functions
df

ID    Name    Income           City   Age Department   Sales Response
0    101   James   50000.0        Chicago  27.0         HR  700000        Y
1    201     Tom   35000.0       New York  33.0  Analytics  550000        Y
2    301   Starc   20000.0         Dallas  21.0         IT  125000        N
3    401    Phil  100000.0         Dallas  40.0         HR  800000        Y
4    501    Rick       NaN  San Francisco   NaN         IT   75000        Y
5    601   Harry  135000.0          Vegas  44.0  Marketing  900000        N
6    701     Jim   25000.0    Los Angeles  24.0         HR  350000        Y
7    801   Ricky   75000.0        Chicago   NaN  Analytics  450000        N
8    901     Fin   85000.0          Vegas  51.0  Marketing  600000        N
9   1001   Grant   61500.0       New York  39.0         IT   75000        N
10  1002  Cooper   31500.0         Dallas  27.0         IT  250000        N
11  1003   Brand   40500.0  San Francisco  49.0  Marketing   50000        N

##### No. of rows and columns in a DataFrame

In [625]:
# df.shape gives number of rows and columns in a DataFrame (rows, columns)
df.shape

(12, 8)

While preparing data for developing a model it is always important to check if the dataset is unique at primary key level i.e.
there are no duplicate records (until and unless required depending on the use case). In SQL we can use count() and 
count (distinct <primary_key>) functions to check if my dataset is unique or not. 

In Python we can use df.shape function to check the total number of records and df['<primary_key>'].nunique() function to check unique number of records in a DataFrame. let us check this on the above DataFrame

##### Count of unique elements in a column

In [627]:
# ID here is primary key, so getting count of unique IDs in our Dataset
df['ID'].nunique()

12

In [629]:
# We can get this count using one more method
len(df['ID'].unique())

12

##### Unique elements in a  column

df.unique() function gets all the unique values for a particular column in a DataFrame

In [478]:
df['City'].unique()

array(['Chicago', 'New York', 'Dallas', 'San Francisco', 'Vegas',
       'Los Angeles'], dtype=object)

City column has 6 unique cities which is the output of unique() function 

##### Count of unique elements in a column

In [485]:
df['City'].nunique()

6

In [486]:
# We can get this count using one more method
len(df['City'].unique())

6

##### Frequency distribution or count of each element in a column

In [491]:
df['City'].value_counts()

Dallas           3
Vegas            2
Chicago          2
San Francisco    2
New York         2
Los Angeles      1
Name: City, dtype: int64

##### Sorting

In [502]:
# By default sort_values will sort in ascending order by the column you have specified
df.sort_values('Income')

ID    Name    Income           City   Age Department   Sales Response
2    301   Starc   20000.0         Dallas  21.0         IT  125000        N
6    701     Jim   25000.0    Los Angeles  24.0         HR  350000        Y
10  1002  Cooper   31500.0         Dallas  27.0         IT  250000        N
1    201     Tom   35000.0       New York  33.0  Analytics  550000        Y
11  1003   Brand   40500.0  San Francisco  49.0  Marketing   50000        N
0    101   James   50000.0        Chicago  27.0         HR  700000        Y
9   1001   Grant   61500.0       New York  39.0         IT   75000        N
7    801   Ricky   75000.0        Chicago   NaN  Analytics  450000        N
8    901     Fin   85000.0          Vegas  51.0  Marketing  600000        N
3    401    Phil  100000.0         Dallas  40.0         HR  800000        Y
5    601   Harry  135000.0          Vegas  44.0  Marketing  900000        N
4    501    Rick       NaN  San Francisco   NaN         IT   75000        Y

In [503]:
# Sorting in descending order
df.sort_values('Income', ascending = False)

ID    Name    Income           City   Age Department   Sales Response
5    601   Harry  135000.0          Vegas  44.0  Marketing  900000        N
3    401    Phil  100000.0         Dallas  40.0         HR  800000        Y
8    901     Fin   85000.0          Vegas  51.0  Marketing  600000        N
7    801   Ricky   75000.0        Chicago   NaN  Analytics  450000        N
9   1001   Grant   61500.0       New York  39.0         IT   75000        N
0    101   James   50000.0        Chicago  27.0         HR  700000        Y
11  1003   Brand   40500.0  San Francisco  49.0  Marketing   50000        N
1    201     Tom   35000.0       New York  33.0  Analytics  550000        Y
10  1002  Cooper   31500.0         Dallas  27.0         IT  250000        N
6    701     Jim   25000.0    Los Angeles  24.0         HR  350000        Y
2    301   Starc   20000.0         Dallas  21.0         IT  125000        N
4    501    Rick       NaN  San Francisco   NaN         IT   75000        Y

In [505]:
# Sorting in descending order by two clumns. This is similar to order by in sql
df.sort_values(['City', 'Sales'], ascending = False)

ID    Name    Income           City   Age Department   Sales Response
5    601   Harry  135000.0          Vegas  44.0  Marketing  900000        N
8    901     Fin   85000.0          Vegas  51.0  Marketing  600000        N
4    501    Rick       NaN  San Francisco   NaN         IT   75000        Y
11  1003   Brand   40500.0  San Francisco  49.0  Marketing   50000        N
1    201     Tom   35000.0       New York  33.0  Analytics  550000        Y
9   1001   Grant   61500.0       New York  39.0         IT   75000        N
6    701     Jim   25000.0    Los Angeles  24.0         HR  350000        Y
3    401    Phil  100000.0         Dallas  40.0         HR  800000        Y
10  1002  Cooper   31500.0         Dallas  27.0         IT  250000        N
2    301   Starc   20000.0         Dallas  21.0         IT  125000        N
0    101   James   50000.0        Chicago  27.0         HR  700000        Y
7    801   Ricky   75000.0        Chicago   NaN  Analytics  450000        N

Indexes will remain attached to as that of original DataFrame